checks.d/kafka_consumer.py

```diff
@@ -29,38 +29,92 @@ def __init__(self, name, init_config, agentConfig, instances=None):
             consumer_offsets = {}
-            topics = defaultdict(set)
-            for consumer_group, topic_partitions in consumer_groups.iteritems():
-                for topic, partitions in topic_partitions.iteritems():
-                    # Remember the topic partitions that we've see so that we can
-                    # look up their broker offsets later
-                    topics[topic].update(set(partitions))
+            # Specifying consumer_groups is optional, if they don't exist, then fetch from ZK
+            if not consumer_groups_specified:
+                consumer_groups = {consumer_group: None for consumer_group in
+                    self._get_children(zk_conn, zk_path_consumer, 'consumer groups')}
+
+            for consumer_group, topics in consumer_groups.iteritems():
+                # Specifying topics is optional, if they don't exist, then fetch from ZK
+                if topics is None:
+                    zk_path_topics = zk_path_topic_tmpl % (consumer_group)
+                    topics = {topic: None for topic in
+                        self._get_children(zk_conn, zk_path_topics, 'topics')}
+
+                for topic, partitions in topics.iteritems():
+                    # Specifying partitions is optional, if they don't exist, then fetch from ZK
+                    if partitions is None:
+                        zk_path_partitions = zk_path_partition_tmpl % (consumer_group, topic)
+                        # Zookeeper returns the partition IDs as strings because
+                        # they are extracted from the node path
+                        partitions = [int(x) for x in self._get_children(
+                            zk_conn, zk_path_partitions, 'partitions')]
+
+                    # Remember these topic partitions so that we can lookup
+                    # the broker offset in the Kafka cluster to calculate
+                    # consumer lag. This topic may already have partitions
+                    # in tracked_topic_partitions from another consumer group,
+                    # so use a set update() to only add missing partitions
+                    tracked_topic_partitions[topic].update(set(partitions))
+
+                    # Fetch consumer offsets for each partition from ZK
                     for partition in partitions:
-                        zk_path = zk_path_tmpl % (consumer_group, topic, partition)
+                        zk_path = (zk_path_partition_tmpl + '%s/') % (consumer_group, topic, partition)
                         try:
                             consumer_offset = int(zk_conn.get(zk_path)[0])
                             key = (consumer_group, topic, partition)
                             consumer_offsets[key] = consumer_offset
                         except NoNodeError:
                             self.log.warn('No zookeeper node at %s' % zk_path)
-                        except Exception:
+                        except:
```

Comment:
Yeah, I'm totally familiar with the rationale, just missed it when cleaning up #2781